In [1]:
list_hot_encoder = ['CreditScore', 'Age', 'Tenure_Months', 'Balance', 'NumOfProducts',
       'EstimatedSalary', 'Loyalty_Score', 'Exited', 'Gender_Female',
       'Gender_Male', 'HasCrCard_No', 'HasCrCard_Yes', 'IsActiveMember_No',
       'IsActiveMember_Yes', 'Credit_Score_Category_Excellent',
       'Credit_Score_Category_Fair', 'Credit_Score_Category_Good',
       'Tenure_Group_0-1 year', 'Tenure_Group_1-3 years',
       'Tenure_Group_3-5 years', 'Tenure_Group_5+ years']

list_label_encoder = ['CreditScore', 'Age', 'Tenure_Months', 'Balance', 'NumOfProducts',
       'EstimatedSalary', 'Loyalty_Score', 'Exited', 'Geography_num', 'Gender_num', 'HasCrCard_num', 'IsActiveMember_num',
       'Credit_Score_Category_num', 'Tenure_Group_num']

list_numeric = ['CreditScore', 'Age', 'Tenure_Months', 'Balance', 'NumOfProducts',
       'EstimatedSalary', 'Loyalty_Score', 'Exited']

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv('/content/drive/MyDrive/treino/classificação de rotatividade de clientes/dados_tratados.csv')
df.head()

CreditScore  Age  Tenure_Months    Balance  NumOfProducts  EstimatedSalary  \
0          619   42             24       0.00              1        101348.88   
1          608   41             12   83807.86              1        112542.58   
2          502   42             96  159660.80              3        113931.57   
3          699   39             12       0.00              2         93826.63   
4          850   43             24  125510.82              1         79084.10   

   Loyalty_Score  Exited Geography  Gender  ... HasCrCard_Yes  \
0              2       1    France  Female  ...           1.0   
1              1       0     Spain  Female  ...           0.0   
2             24       1    France  Female  ...           1.0   
3              2       0    France  Female  ...           0.0   
4              2       0     Spain  Female  ...           1.0   

  IsActiveMember_No IsActiveMember_Yes Credit_Score_Category_Excellent  \
0               0.0                1.0                             0.0   
1               0.0                1.0                             0.0   
2               1.0                0.0                             0.0   
3               1.0                0.0                             0.0   
4               0.0                1.0                             1.0   

   Credit_Score_Category_Fair  Credit_Score_Category_Good  \
0                         0.0                         1.0   
1                         0.0                         1.0   
2                         1.0                         0.0   
3                         0.0                         1.0   
4                         0.0                         0.0   

   Tenure_Group_0-1 year  Tenure_Group_1-3 years  Tenure_Group_3-5 years  \
0                    0.0                     1.0                     0.0   
1                    1.0                     0.0                     0.0   
2                    0.0                     0.0                     0.0   
3                    1.0                     0.0                     0.0   
4                    0.0                     1.0                     0.0   

   Tenure_Group_5+ years  
0                    0.0  
1                    0.0  
2                    1.0  
3                    0.0  
4                    0.0  

[5 rows x 33 columns]

In [4]:
from sklearn.svm import SVC
from sklearn.linear_model import SGDClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.model_selection import train_test_split

In [5]:
def mostrar_acuracia(modelo, X_train, X_test, y_train, y_test):
  y_pred = modelo.predict(X_test)
  y_pred_train = modelo.predict(X_train)
  print('='*25)
  print('Teste')
  print('='*25)
  print(accuracy_score(y_test, y_pred))
  print(confusion_matrix(y_test, y_pred))
  print('='*25)
  print('Treino')
  print('='*25)
  print(accuracy_score(y_train, y_pred_train))
  print(confusion_matrix(y_train, y_pred_train))

def sgd(X_train, X_test, y_train, y_test):
  sgd = SGDClassifier()
  sgd.fit(X_train, y_train)
  mostrar_acuracia(sgd, X_train, X_test, y_train, y_test)

def svc(X_train, X_test, y_train, y_test):
  svc = SVC()
  svc.fit(X_train, y_train)
  mostrar_acuracia(svc, X_train, X_test, y_train, y_test)

def knn(X_train, X_test, y_train, y_test, n_neighbors=5):
  knn = KNeighborsClassifier(n_neighbors=n_neighbors)
  knn.fit(X_train, y_train)
  mostrar_acuracia(knn, X_train, X_test, y_train, y_test)

def gausiano(X_train, X_test, y_train, y_test):
  gausiano = GaussianProcessClassifier()
  gausiano.fit(X_train, y_train)
  mostrar_acuracia(gausiano, X_train, X_test, y_train, y_test)

def arvore(X_train, X_test, y_train, y_test, criterion='gini', splitter='best', max_depth=None):
  arvore = DecisionTreeClassifier(criterion=criterion, splitter=splitter, max_depth=max_depth, random_state=42)
  arvore.fit(X_train, y_train)
  mostrar_acuracia(arvore, X_train, X_test, y_train, y_test)

def floresta(X_train, X_test, y_train, y_test, n_estimators=100, criterion='gini', max_depth=None):
  floresta = RandomForestClassifier(n_estimators=n_estimators, criterion=criterion, max_depth=max_depth, random_state=42)
  floresta.fit(X_train, y_train)
  mostrar_acuracia(floresta, X_train, X_test, y_train, y_test)

Tabela com apenas valores numéricos

In [6]:
df1 = df[list_numeric]
df1.head(1)

CreditScore  Age  Tenure_Months  Balance  NumOfProducts  EstimatedSalary  \
0          619   42             24      0.0              1        101348.88   

   Loyalty_Score  Exited  
0              2       1

In [7]:
from sklearn.model_selection import train_test_split

In [8]:
X = df1.drop('Exited', axis=1)
y = df1['Exited']

In [9]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [10]:
sgd(X_train, X_test, y_train, y_test)

Teste
0.8053333333333333
[[2416    0]
 [ 584    0]]
Treino
0.7924285714285715
[[5547    0]
 [1453    0]]


In [11]:
svc(X_train, X_test, y_train, y_test)

Teste
0.8053333333333333
[[2416    0]
 [ 584    0]]
Treino
0.7924285714285715
[[5547    0]
 [1453    0]]


In [12]:
knn(X_train, X_test, y_train, y_test, n_neighbors=10)

Teste
0.801
[[2392   24]
 [ 573   11]]
Treino
0.7957142857142857
[[5502   45]
 [1385   68]]


In [13]:
arvore(X_train, X_test, y_train, y_test, criterion='gini', splitter='best', max_depth=4)

Teste
0.849
[[2270  146]
 [ 307  277]]
Treino
0.8384285714285714
[[5177  370]
 [ 761  692]]


In [14]:
floresta(X_train, X_test, y_train, y_test, n_estimators=40, criterion='gini', max_depth=9)

Teste
0.8516666666666667
[[2328   88]
 [ 357  227]]
Treino
0.8712857142857143
[[5428  119]
 [ 782  671]]


DataFrame com valores label encoder

In [15]:
df2 = df[list_label_encoder]
df2.head(1)

CreditScore  Age  Tenure_Months  Balance  NumOfProducts  EstimatedSalary  \
0          619   42             24      0.0              1        101348.88   

   Loyalty_Score  Exited  Geography_num  Gender_num  HasCrCard_num  \
0              2       1              0           0              1   

   IsActiveMember_num  Credit_Score_Category_num  Tenure_Group_num  
0                   1                          2                 1

In [16]:
X = df2.drop('Exited', axis=1)
y = df2['Exited']

In [17]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [18]:
sgd(X_train, X_test, y_train, y_test)

Teste
0.7326666666666667
[[2081  335]
 [ 467  117]]
Treino
0.7217142857142858
[[4811  736]
 [1212  241]]


In [19]:
svc(X_train, X_test, y_train, y_test)

Teste
0.8053333333333333
[[2416    0]
 [ 584    0]]
Treino
0.7924285714285715
[[5547    0]
 [1453    0]]


In [20]:
knn(X_train, X_test, y_train, y_test, n_neighbors=10)

Teste
0.801
[[2392   24]
 [ 573   11]]
Treino
0.7957142857142857
[[5502   45]
 [1385   68]]


In [21]:
arvore(X_train, X_test, y_train, y_test, criterion='gini', splitter='best', max_depth=6)

Teste
0.861
[[2358   58]
 [ 359  225]]
Treino
0.8612857142857143
[[5421  126]
 [ 845  608]]


In [22]:
floresta(X_train, X_test, y_train, y_test, n_estimators=200, criterion='gini', max_depth=11)

Teste
0.871
[[2363   53]
 [ 334  250]]
Treino
0.9085714285714286
[[5513   34]
 [ 606  847]]


Valores Hot Encoder

In [23]:
df3 = df[list_hot_encoder]
df3.head(1)

CreditScore  Age  Tenure_Months  Balance  NumOfProducts  EstimatedSalary  \
0          619   42             24      0.0              1        101348.88   

   Loyalty_Score  Exited  Gender_Female  Gender_Male  ...  HasCrCard_Yes  \
0              2       1            1.0          0.0  ...            1.0   

   IsActiveMember_No  IsActiveMember_Yes  Credit_Score_Category_Excellent  \
0                0.0                 1.0                              0.0   

   Credit_Score_Category_Fair  Credit_Score_Category_Good  \
0                         0.0                         1.0   

   Tenure_Group_0-1 year  Tenure_Group_1-3 years  Tenure_Group_3-5 years  \
0                    0.0                     1.0                     0.0   

   Tenure_Group_5+ years  
0                    0.0  

[1 rows x 21 columns]

In [24]:
X = df3.drop('Exited', axis=1)
y = df['Exited']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [25]:
sgd(X_train, X_test, y_train, y_test)

Teste
0.224
[[  99 2317]
 [  11  573]]
Treino
0.2337142857142857
[[ 219 5328]
 [  36 1417]]


In [26]:
svc(X_train, X_test, y_train, y_test)

Teste
0.8053333333333333
[[2416    0]
 [ 584    0]]
Treino
0.7924285714285715
[[5547    0]
 [1453    0]]


In [27]:
knn(X_train, X_test, y_train, y_test, n_neighbors=10)

Teste
0.801
[[2392   24]
 [ 573   11]]
Treino
0.7957142857142857
[[5502   45]
 [1385   68]]


In [28]:
arvore(X_train, X_test, y_train, y_test, criterion='gini', splitter='best', max_depth=6)

Teste
0.8593333333333333
[[2357   59]
 [ 363  221]]
Treino
0.861
[[5418  129]
 [ 844  609]]


In [29]:
floresta(X_train, X_test, y_train, y_test, n_estimators=200, criterion='gini', max_depth=11)

Teste
0.8626666666666667
[[2360   56]
 [ 356  228]]
Treino
0.8967142857142857
[[5513   34]
 [ 689  764]]


In [30]:
from sklearn.preprocessing import MinMaxScaler

In [31]:
scaler = MinMaxScaler()

df_scaled = scaler.fit_transform(df2)

df_normalized = pd.DataFrame(df_scaled, columns=df2.columns)
df_normalized.head(1)

CreditScore       Age  Tenure_Months  Balance  NumOfProducts  \
0        0.538  0.324324            0.2      0.0            0.0   

   EstimatedSalary  Loyalty_Score  Exited  Geography_num  Gender_num  \
0         0.506735           0.05     1.0            0.0         0.0   

   HasCrCard_num  IsActiveMember_num  Credit_Score_Category_num  \
0            1.0                 1.0                        1.0   

   Tenure_Group_num  
0          0.333333

In [32]:
X = df3.drop('Exited', axis=1)
y = df['Exited']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

In [33]:
sgd(X_train, X_test, y_train, y_test)

Teste
0.8053333333333333
[[2416    0]
 [ 584    0]]
Treino
0.7924285714285715
[[5547    0]
 [1453    0]]


In [34]:
svc(X_train, X_test, y_train, y_test)

Teste
0.8053333333333333
[[2416    0]
 [ 584    0]]
Treino
0.7924285714285715
[[5547    0]
 [1453    0]]


In [35]:
knn(X_train, X_test, y_train, y_test, n_neighbors=10)

Teste
0.801
[[2392   24]
 [ 573   11]]
Treino
0.7957142857142857
[[5502   45]
 [1385   68]]


In [36]:
arvore(X_train, X_test, y_train, y_test, criterion='entropy', splitter='best', max_depth=6)

Teste
0.8606666666666667
[[2364   52]
 [ 366  218]]
Treino
0.8582857142857143
[[5428  119]
 [ 873  580]]


In [37]:
floresta(X_train, X_test, y_train, y_test, n_estimators=100, criterion='gini', max_depth=10)

Teste
0.8636666666666667
[[2366   50]
 [ 359  225]]
Treino
0.8832857142857143
[[5495   52]
 [ 765  688]]


In [38]:
X = df2.drop('Exited', axis=1)
y = df2['Exited']

In [39]:
model = RandomForestClassifier(n_estimators=200, criterion='gini', max_depth=11, random_state=42)
model.fit(X, y)

RandomForestClassifier(max_depth=11, n_estimators=200, random_state=42)

In [40]:
import joblib

joblib.dump(model, 'modelo_classificador_de_rotatividade.pkl')

['modelo_classificador_de_rotatividade.pkl']

In [43]:
df2.head(1)

CreditScore  Age  Tenure_Months  Balance  NumOfProducts  EstimatedSalary  \
0          619   42             24      0.0              1        101348.88   

   Loyalty_Score  Exited  Geography_num  Gender_num  HasCrCard_num  \
0              2       1              0           0              1   

   IsActiveMember_num  Credit_Score_Category_num  Tenure_Group_num  
0                   1                          2                 1


* Apenas Valores numéricos: teste = 85%, treino = 87% com floresta(X_train, X_test, y_train, y_test, n_estimators=40, criterion='gini', max_depth=9)

* Valores label encoder: teste = 87%, treino = 90% com floresta(X_train, X_test, y_train, y_test, n_estimators=200, criterion='gini', max_depth=11)

* Valores Hot encoder: teste = 86%, treino = 89% com floresta(X_train, X_test, y_train, y_test, n_estimators=200, criterion='gini', max_depth=11)

* Valores Normalizado: teste = 86,3%, treino = 88,3% com floresta(X_train, X_test, y_train, y_test, n_estimators=200, criterion='gini', max_depth=11)

